pkg/trace/api/otlp.go

```diff
@@ -0,0 +1,428 @@
+		} else {
+			span.Meta["events"] = string(out)
+		}
+	}
+	for _, kv := range in.Attributes {
+		switch v := kv.Value.Value.(type) {
+		case *otlppb.AnyValue_DoubleValue:
+			span.Metrics[kv.Key] = v.DoubleValue
+		case *otlppb.AnyValue_IntValue:
+			span.Metrics[kv.Key] = float64(v.IntValue)
+		default:
+			span.Meta[kv.Key] = anyValueString(kv.Value)
+		}
+	}
+	if _, ok := span.Meta["env"]; !ok {
+		if env := span.Meta["deployment.environment"]; env != "" {
+			span.Meta["env"] = env
+		}
+	}
+	if in.TraceState != "" {
+		span.Meta["trace_state"] = in.TraceState
+	}
+	if lib.Name != "" {
+		span.Meta["instrumentation_library.name"] = lib.Name
+	}
+	if lib.Version != "" {
+		span.Meta["instrumentation_library.version"] = lib.Version
+	}
+	if svc := span.Meta["peer.service"]; svc != "" {
+		span.Service = svc
+	}
+	if r := resourceFromTags(span.Meta); r != "" {
+		span.Resource = r
+	}
+	span.Type = spanKind2Type(in.Kind, span)
+	status2Error(in.Status, in.Events, span)
+	return span
+}
+
+// resourceFromTags attempts to deduce a more accurate span resource from the given list of tags meta.
+// If this is not possible, it returns an empty string.
+func resourceFromTags(meta map[string]string) string {
+	var r string
+	if m := meta["http.method"]; m != "" {
+		r = m
+		if route := meta["http.route"]; route != "" {
+			r += " " + route
+		} else if route := meta["grpc.path"]; route != "" {
+			r += " " + route
+		}
+	} else if m := meta["messaging.operation"]; m != "" {
+		r = m
+		if dest := meta["messaging.destination"]; dest != "" {
+			r += " " + dest
+		}
+	}
+	return r
+}
+
+// status2Error checks the given status and events and applies any potential error and messages
+// to the given span attributes.
+func status2Error(status *otlppb.Status, events []*otlppb.Span_Event, span *pb.Span) {
+	if status == nil || status.Code != otlppb.Status_STATUS_CODE_ERROR {
+		return
+	}
+	span.Error = 1
+	for _, e := range events {
+		if strings.ToLower(e.Name) != "exception" {
+			continue
+		}
+		for _, attr := range e.Attributes {
+			switch attr.Key {
+			case "exception.message":
+				span.Meta["error.msg"] = anyValueString(attr.Value)
+			case "exception.type":
+				span.Meta["error.type"] = anyValueString(attr.Value)
+			case "exception.stacktrace":
+				span.Meta["error.stack"] = anyValueString(attr.Value)
+			}
+		}
+	}
+}
+
+// spanKind2Type returns a span's type based on the given kind and other present properties.
+func spanKind2Type(kind otlppb.Span_SpanKind, span *pb.Span) string {
+	var typ string
+	switch kind {
+	case otlppb.Span_SPAN_KIND_SERVER:
+		typ = "web"
+	case otlppb.Span_SPAN_KIND_CLIENT:
+		typ = "http"
+		db, ok := span.Meta["db.system"]
+		if !ok {
+			break
+		}
+		switch db {
+		case "redis", "memcached":
+			typ = "cache"
+		default:
+			typ = "db"
+		}
+	default:
+		typ = "custom"
+	}
+	return typ
+}
+
+func byteArrayToUint64(b []byte) uint64 {
+	if len(b) < 8 {
+		return 0
+	}
+	return binary.LittleEndian.Uint64(b)
```

Comment:
just wondering is this the same implementation as the one here?

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/f1e8331be111a4a8bf65670dfaff359714edef85/exporter/datadogexporter/translate_traces.go#L465

it looks different but im not a pro in golang. might be best to align 

https://docs.datadoghq.com/tracing/connect_logs_and_traces/opentelemetry/

Reply:
Not entirely the same, I think. I was wondering about this myself, thanks for pointing me to the link. I will update this.

Reply:
I ran some tests ([link](https://play.golang.org/p/Wm9yZagS54z)) and updated the algorithm to match the output, in the latest commit. PTAL

Reply:
I'm not an entirely sure of this I'm not an employee of Datadog 😅 maybe @ericmustin can confirm

Reply:
It's fine. It's the same thing. Thanks for the comment.